### DISCUSSION TITLE
Django integration Testing

### DISCUSSION THREAD

**Discussion by slayb3ar** [procrastinate-org/procrastinate#1042] [Q&A] [answered]
Have recently integrated procrastinate as a basic job queue system in a django project. Has been a pleasure so far 🙌
Am now working on adding some test coverage as well. I'm following the guide here: https://procrastinate.readthedocs.io/en/stable/howto/django/tests.html

What's unclear to me - can one actually test task  _execution_, and not just creation?
I can confirm my task is being created, with a status of `todo`, but nothing beyond that. 

Is there a way to run a worker within an integration test?

More context: Using Django's `TestCase` ++ `manage.py test`

**Comment by ewjoachim** [comment]
Yes, you can run a worker in your integration tests, e.g. through [`app.run_worker(wait=False, install_signal_handlers=False)`](https://procrastinate.readthedocs.io/en/stable/reference.html#procrastinate.App.run_worker).

- `wait=False` means that when all tasks are executed, the call will return.
- `install_signal_handlers=False` is optional and just here to keep the worker from changing the signal callbacks set by your test runner.

**Comment by slayb3ar** [comment]
Thanks @ewjoachim! It seems to be working fine locally. One note for anyone who runs into this: I believe you need to use `TransactionTestCase` over `TestCase` here, as the latter will fail silently on `select_for_update` queries. See: https://stackoverflow.com/questions/44450533/difference-between-testcase-and-transactiontestcase-classes-in-django-test

Below is working example 
```
from procrastinate.contrib.django import app
from django.test import TransactionTestCase

class TestEventDetailViews(TransactionTestCase):
	def testCase(self):
		# Run tasks
		...
		
		# Start worker
		app = app.with_connector(app.connector.get_worker_connector())
		app.run_worker(wait=False, install_signal_handlers=False, listen_notify=True)
```

**Reply by ewjoachim** [reply]
Good! I tend to use pytest much more, but it's something that could be written in the test part of django section of the procrastinate doc. Would you like to add it ?

**Reply by slayb3ar** [reply, marked as answer]
yes definitely! I just opened up a PR here: 
https://github.com/procrastinate-org/procrastinate/pull/1046
